spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -168,11 +194,39 @@ public long count(Class<?> entityClass) {
 
 	@Override
 	public <T> Collection<T> findAllById(Iterable<?> ids, Class<T> entityClass) {
-		List<Key> keysToFind = getKeysFromIds(ids, entityClass);
-		return convertEntitiesForRead(
-				getDatastoreReadWriter().get(
-						keysToFind.toArray(new Key[keysToFind.size()])),
-				entityClass);
+		return findAllById(ids, entityClass, new ReadContext());
+	}
+
+	private <T> Collection<T> findAllById(Iterable<?> ids, Class<T> entityClass, ReadContext context) {
+		List<Key> keys = StreamSupport.stream(ids.spliterator(), false).map((id) -> getKeyFromId(id, entityClass))
+				.collect(Collectors.toList());
+		List<Key> missingKeys = keys.stream().filter(context::notCached).collect(Collectors.toList());
+
+		if (!missingKeys.isEmpty()) {
+			List<Entity> entities = getDatastoreReadWriter().fetch(missingKeys.toArray(new Key[] {}));
+
+			for (int i = 0; i < missingKeys.size(); i++) {
+				Key key = missingKeys.get(i);
+				context.putReadEntity(key, entities.get(i));
+			}
+		}
+
+		ArrayList<T> result = new ArrayList<>();
+		for (Key key : keys) {
+			T convertedEntity;
+			if (context.converted(key)) {
+				convertedEntity = (T) context.getConvertedEntity(key);
+			}
+			else {
+				Object entity = context.getReadEntity(key);
+				convertedEntity = convertEntitiesForRead(Collections.singletonList((BaseEntity) entity).iterator(), entityClass,
```

Comment:
Because it doesn't look like you're adding to the context in each iteration of this for-loop, I don't think you need to create singleton lists for convertEntitiesForRead. The loop inside convertEntitiesForRead looks like it will work just fine if you gave it all the entities at the same time.